torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,289 @@
+      else:
+        logging.info("XLA Backend: Skipping non-tensor wait.")
+
+
+# -------------------------------------
+# Override torch.distributed.new_group.
+# -------------------------------------
+_orig_new_group_fn = dist.new_group
+
+
+def infer_mesh(slice_ranks, world_size):
+  '''Infers a rectangular mesh topology from a slice in the mesh.
+
+    Example, given world size 12 and a slice like the following:
+        [1, 5, 9]
+    this function infers the following mesh:
+        [[0, 4, 8],
+         [1, 5, 9],
+         [2, 6, 10],
+         [3, 7, 11]]
+
+    We only support ractangular meshes.
+    '''
+  slice_len = len(slice_ranks)
+  if world_size % slice_len != 0:
+    raise ValueError(
+        'Given slice length doesn\'t equal to a side of a rectangle. '
+        f'World size: {world_size}, slice length: {slice_len}')
+
+  # ensure the ranks are a correct range.
+  start = slice_ranks[0]
+  step = slice_ranks[1] - slice_ranks[0]
+  stop = start + step * slice_len
+  expected_ranks = list(range(start, stop, step))
+  if slice_ranks != expected_ranks:
+    raise ValueError(
+        f'Given slice isn\'t a range with evenly distributed steps: {slice_ranks}'
+    )
+
+  num_slices = int(world_size / slice_len)
+  world = list(range(world_size))
+  mesh = []
+  if step == 1:
+    # Horizontal case.
+    if start % slice_len != 0 or start >= world_size:
+      raise ValueError('Given horizontal slice doesn\'t have a correct start: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    for i in range(num_slices):
+      slice_start = i * slice_len
+      slice_stop = (i + 1) * slice_len
+      slice_idx = slice(slice_start, slice_stop, step)
+      mesh.append(world[slice_idx])
+  else:
+    # Vertical case.
+    if start >= num_slices:
+      raise ValueError('Given vertical slice doesn\'t have a correct start: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    if step != num_slices:
+      raise ValueError('Given vertical slice doesn\'t have a correct step: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    for i in range(num_slices):
+      slice_start = i
+      slice_stop = i + slice_len * step
+      slice_idx = slice(slice_start, slice_stop, step)
+      mesh.append(world[slice_idx])
+  assert slice_ranks in mesh
+
+  return mesh
+
+
+def new_xla_process_group(ranks=None,
+                          timeout=dist.default_pg_timeout,
+                          backend=None):
+  pg = _orig_new_group_fn(ranks, timeout, backend)
+  if isinstance(pg, ProcessGroupXla) and ranks is not None:
+    world_pg = dist.group.WORLD
+    if not isinstance(world_pg, ProcessGroupXla):
+      raise RuntimeError('xla backend requires the default ProcessGroup to be '
```

Comment:
Is this a hard requirement or just a temporary solution?

Reply:
We didn't plan or test the scenario where we have a non-XLA process group as the world and other XLA process groups working with it. Is this a legit use case? Doing that will at least cause data dependency knowledge being incomplete in the XLA graphs.

Reply:
not a common use case, but I have seen users creating Gloo backend for CPU communication (say sending input data from dataloader to trainer) and NCCL backend for GPU communication (gradient synchronization). The behavior should be good for now. We can always make it more flexible when users ask.

Reply:
ACK